docs/type_inference.md

```diff
@@ -0,0 +1,60 @@
+# Type Inference
+
+StableHLO is an operation set that expresses ML computations. It has been originally bootstrapped from [the MHLO dialect](https://github.com/tensorflow/mlir-hlo#meta-hlo-dialect-mhlo), including inheriting the implementation of type inference. The implementation progress is tracked in [status.md](https://github.com/openxla/stablehlo/blob/main/docs/status.md). 
+
+To implement high-quality verifiers and shape functions for StableHLO ops, these guidelines and criterion are proposed below to follow:
+
+# Proposal
+
+These proposals apply to both revisiting existing implementations, and achieving new ops til a comprehensive coverage.
+
+## (P1) Use the StableHLO spec as the source of truth. 
+
+The [spec](https://github.com/openxla/stablehlo/blob/main/docs/spec_draft.md) is the source of truth for all verifications of the StableHLO ops. The existing type inference implementation of every ops need revisited to be fully aligned with the specification. Note that the specification document keeps evolving, in cases that the spec for an op is not available, the XLA implementation should be used as the source of truth instead: including [xla/service/shape\_inference.cc](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/service/shape_inference.cc) and [xla/service/hlo\_verifier.cc](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/service/hlo_verifier.cc). XLA implementation doesn't cover unbounded dynamism, so for unbounded dynamism we'll apply common sense until the dynamism RFC is available.
```

Comment:
Also, "The existing verifiers and shape functions need to be revisited".

Reply:
Done.